File: tests/java/src/test/java/com/aptos/specs/steps/AddressSteps.java
package com.aptos.specs.steps;

import com.aptos.specs.support.World;
import com.aptos.specs.support.Vectors;
import com.aptoslabs.japtos.types.AccountAddress;
import com.aptoslabs.japtos.utils.HexUtils;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.And;

import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.*;

/**
 * Step definitions for address.feature
 * 
 * These steps test AccountAddress parsing, formatting, comparison, and BCS serialization
 * using the japtos SDK.
 */
public class AddressSteps {
    
    private final World world;
    
    public AddressSteps(World world) {
        this.world = world;
    }
    
    // ==========================================================================
    // Given Steps
    // ==========================================================================
    
    @Given("a hex string {string}")
    public void givenHexString(String hexString) {
        world.setHexString(hexString);
    }
    
    @Given("an AccountAddress with value {int}")
    public void givenAccountAddressWithValue(int value) {
        try {
            // Create address from hex representation of the value
            String hex = String.format("0x%x", value);
            AccountAddress addr = AccountAddress.fromHex(hex);
            world.setAddress(addr);
            world.setBytes(addr.toBytes());
            world.clearError();
        } catch (Exception e) {
            world.setError(e);
        }
    }
    
    @Given("an AccountAddress from hex {string}")
    public void givenAccountAddressFromHex(String hexString) {
        try {
            AccountAddress addr = AccountAddress.fromHex(hexString);
            world.setAddress(addr);
            world.clearError();
        } catch (Exception e) {
            world.setError(e);
        }
    }
    
    @Given("another AccountAddress from hex {string}")
    public void givenAnotherAccountAddressFromHex(String hexString) {
        try {
            AccountAddress addr = AccountAddress.fromHex(hexString);
            world.setAddress2(addr);
            world.clearError();
        } catch (Exception e) {
            world.setError(e);
        }
    }
    
    @Given("the ZERO address constant")
    public void givenZeroAddressConstant() {
        world.setAddress(AccountAddress.ZERO);
    }
    
    @Given("the ONE address constant")
    public void givenOneAddressConstant() {
        world.setAddress(AccountAddress.ONE);
    }
    
    @Given("the THREE address constant")
    public void givenThreeAddressConstant() {
        world.setAddress(AccountAddress.THREE);
    }
    
    @Given("the FOUR address constant")
    public void givenFourAddressConstant() {
        world.setAddress(AccountAddress.FOUR);
    }
    
    @Given("{int} bytes with value {int} in the last byte")
    public void givenBytesWithValueInLastByte(int numBytes, int value) {
        byte[] bytes = new byte[numBytes];
        bytes[numBytes - 1] = (byte) value;
        world.setBytes(bytes);
    }
    
    @Given("test vectors from addresses.json")
    public void givenAddressTestVectors() throws Exception {
        List<Vectors.AddressVector> vectors = Vectors.getAddressParsingVectors();
        world.getTestVectors().put("address_parsing", vectors);
    }
    
    // ==========================================================================
    // When Steps
    // ==========================================================================
    
    @When("I parse it as an AccountAddress")
    public void whenParseAsAccountAddress() {
        try {
            AccountAddress addr = AccountAddress.fromHex(world.getHexString());
            world.setAddress(addr);
            world.setBytes(addr.toBytes());
            world.clearError();
        } catch (Exception e) {
            world.setError(e);
        }
    }
    
    @When("I format it as full hex")
    public void whenFormatAsFullHex() {
        AccountAddress addr = (AccountAddress) world.getAddress();
        world.setResult(addr.toHexStringLong());
    }
    
    @When("I format it as short string")
    public void whenFormatAsShortString() {
        AccountAddress addr = (AccountAddress) world.getAddress();
        world.setResult(addr.toHexStringShort());
    }
    
    @When("I BCS serialize the address")
    public void whenBcsSerializeAddress() {
        AccountAddress addr = (AccountAddress) world.getAddress();
        byte[] bcsBytes = addr.toBytes(); // BCS serialization of address is just the 32 bytes
        world.setSerializedBytes(bcsBytes);
        world.setBytes(bcsBytes);
    }
    
    @When("I BCS deserialize as AccountAddress")
    public void whenBcsDeserializeAsAccountAddress() {
        try {
            byte[] bytes = world.getBytes();
            AccountAddress addr = AccountAddress.fromBytes(bytes);
            world.setAddress(addr);
            world.clearError();
        } catch (Exception e) {
            world.setError(e);
        }
    }
    
    @When("I BCS deserialize the result as AccountAddress")
    public void whenBcsDeserializeResultAsAccountAddress() {
        try {
            byte[] bytes = world.getSerializedBytes();
            AccountAddress addr = AccountAddress.fromBytes(bytes);
            world.setResult(addr);
            world.clearError();
        } catch (Exception e) {
            world.setError(e);
        }
    }
    
    @When("I run all parsing test vectors")
    public void whenRunAllParsingTestVectors() {
        @SuppressWarnings("unchecked")
        List<Vectors.AddressVector> vectors = 
            (List<Vectors.AddressVector>) world.getTestVectors().get("address_parsing");
        
        List<TestResult> results = new java.util.ArrayList<>();
        for (Vectors.AddressVector v : vectors) {
            try {
                AccountAddress addr = AccountAddress.fromHex(v.input);
                String fullHex = addr.toHexStringLong();
                String shortString = addr.toHexStringShort();
                
                boolean passed = fullHex.equalsIgnoreCase(v.expected.full_hex) &&
                                 shortString.equalsIgnoreCase(v.expected.short_string);
                
                results.add(new TestResult(v.name, passed, null));
            } catch (Exception e) {
                results.add(new TestResult(v.name, false, e.getMessage()));
            }
        }
        world.setResult(results);
    }
    
    // ==========================================================================
    // Then Steps
    // ==========================================================================
    
    @Then("the parsing should succeed")
    public void thenParsingShouldSucceed() {
        assertThat(world.getError())
            .as("Expected parsing to succeed, but got error: %s", 
                world.getError() != null ? world.getError().getMessage() : "")
            .isNull();
        assertThat(world.getAddress()).isNotNull();
    }
    
    @Then("the parsing should fail with an invalid address error")
    public void thenParsingShouldFailWithInvalidAddressError() {
        assertThat(world.getError())
            .as("Expected parsing to fail with invalid address error")
            .isNotNull();
    }
    
    @Then("the parsing should fail with an invalid hex error")
    public void thenParsingShouldFailWithInvalidHexError() {
        assertThat(world.getError())
            .as("Expected parsing to fail with invalid hex error")
            .isNotNull();
    }
    
    @Then("the parsing should fail with an invalid length error")
    public void thenParsingShouldFailWithInvalidLengthError() {
        assertThat(world.getError())
            .as("Expected parsing to fail with invalid length error")
            .isNotNull();
    }
    
    @Then("the address bytes should have length {int}")
    public void thenAddressBytesHaveLength(int expectedLength) {
        assertThat(world.getBytes()).hasSize(expectedLength);
    }
    
    @Then("byte {int} should equal {int}")
    public void thenByteAtIndexEquals(int index, int value) {
        assertThat(world.getBytes()[index] & 0xFF).isEqualTo(value);
    }
    
    @Then("bytes {int}-{int} should all be {int}")
    public void thenBytesInRangeShouldBe(int start, int end, int value) {
        for (int i = start; i <= end; i++) {
            assertThat(world.getBytes()[i] & 0xFF)
                .as("Byte at index %d", i)
                .isEqualTo(value);
        }
    }
    
    @Then("all {int} bytes should be {int}")
    public void thenAllBytesShouldBe(int count, int value) {
        assertThat(world.getBytes()).hasSize(count);
        for (int i = 0; i < count; i++) {
            assertThat(world.getBytes()[i] & 0xFF)
                .as("Byte at index %d", i)
                .isEqualTo(value);
        }
    }
    
    @Then("the short string should be {string}")
    public void thenShortStringShouldBe(String expected) {
        AccountAddress addr = (AccountAddress) world.getAddress();
        assertThat(addr.toHexStringShort()).isEqualToIgnoringCase(expected);
    }
    
    @Then("the full hex should be {string}")
    public void thenFullHexShouldBe(String expected) {
        AccountAddress addr = (AccountAddress) world.getAddress();
        assertThat(addr.toHexStringLong()).isEqualToIgnoringCase(expected);
    }
    
    @Then("the result should be {string}")
    public void thenResultShouldBe(String expected) {
        assertThat(world.getResult().toString()).isEqualToIgnoringCase(expected);
    }
    
    @Then("the result should be {int} bytes")
    public void thenResultShouldBeBytes(int expectedLength) {
        assertThat(world.getSerializedBytes()).hasSize(expectedLength);
    }
    
    @Then("the two addresses should be equal")
    public void thenTwoAddressesShouldBeEqual() {
        AccountAddress addr1 = (AccountAddress) world.getAddress();
        AccountAddress addr2 = (AccountAddress) world.getAddress2();
        assertThat(addr1).isEqualTo(addr2);
    }
    
    @Then("the two addresses should not be equal")
    public void thenTwoAddressesShouldNotBeEqual() {
        AccountAddress addr1 = (AccountAddress) world.getAddress();
        AccountAddress addr2 = (AccountAddress) world.getAddress2();
        assertThat(addr1).isNotEqualTo(addr2);
    }
    
    @Then("the result should equal the original address")
    public void thenResultShouldEqualOriginalAddress() {
        AccountAddress original = (AccountAddress) world.getAddress();
        AccountAddress result = (AccountAddress) world.getResult();
        assertThat(result).isEqualTo(original);
    }
    
    @Then("all test vectors should pass")
    public void thenAllTestVectorsShouldPass() {
        @SuppressWarnings("unchecked")
        List<TestResult> results = (List<TestResult>) world.getResult();
        
        List<TestResult> failures = results.stream()
            .filter(r -> !r.passed)
            .toList();
        
        assertThat(failures)
            .as("Failed test vectors: %s", 
                failures.stream().map(f -> f.name + ": " + f.error).toList())
            .isEmpty();
    }
    
    // ==========================================================================
    // Helper Classes
    // ==========================================================================
    
    /**
     * Simple test result holder.
     */
    private static class TestResult {
        final String name;
        final boolean passed;
        final String error;
        
        TestResult(String name, boolean passed, String error) {
            this.name = name;
            this.passed = passed;
            this.error = error;
        }
    }
}

File: tests/kotlin/src/test/kotlin/com/aptos/specs/steps/AddressSteps.kt
package com.aptos.specs.steps

import com.aptos.specs.support.World
import com.aptos.specs.support.hexToBytes
import com.aptos.specs.support.toHex
import io.cucumber.java.en.Given
import io.cucumber.java.en.When
import io.cucumber.java.en.Then
import io.cucumber.java.en.And
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import xyz.mcxross.kaptos.model.AccountAddress

/**
 * Step definitions for address parsing and formatting scenarios.
 * 
 * Feature: features/01-core-types/address.feature
 * 
 * Note: Uses Kaptos SDK 0.1.2-beta. Some methods may need adjustment
 * as the SDK API evolves.
 */
class AddressSteps(private val world: World) {
    
    // Helper to get address bytes
    private fun getAddressBytes(): ByteArray {
        val addr = world.address as AccountAddress
        return addr.toString().removePrefix("0x").padStart(64, '0').hexToBytes()
    }
    
    // ============================================================
    // Given Steps
    // ============================================================
    
    @Given("a hex string {string}")
    fun givenAHexString(hex: String) {
        world.hexString = hex
    }
    
    @Given("a short address {string}")
    fun givenAShortAddress(address: String) {
        world.hexString = address
    }
    
    @Given("a full address {string}")
    fun givenAFullAddress(address: String) {
        world.hexString = address
    }
    
    @Given("an address from bytes")
    fun givenAnAddressFromBytes() {
        world.bytes shouldNotBe null
    }
    
    @Given("an AccountAddress with value {int}")
    fun givenAnAccountAddressWithValue(value: Int) {
        runCatching {
            world.address = AccountAddress.fromString("0x${value.toString(16)}")
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @Given("an AccountAddress from hex {string}")
    fun givenAnAccountAddressFromHex(hex: String) {
        runCatching {
            world.address = AccountAddress.fromString(hex)
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @Given("another AccountAddress from hex {string}")
    fun givenAnotherAccountAddressFromHex(hex: String) {
        runCatching {
            world.address2 = AccountAddress.fromString(hex)
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @Given("{int} bytes with value {int} in the last byte")
    fun givenBytesWithValueInLastByte(size: Int, value: Int) {
        val bytes = ByteArray(size)
        bytes[size - 1] = value.toByte()
        world.bytes = bytes
    }
    
    // ============================================================
    // When Steps
    // ============================================================
    
    @When("I parse it as an AccountAddress")
    fun whenIParseItAsAnAccountAddress() {
        runCatching {
            world.address = AccountAddress.fromString(world.hexString!!)
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I format the address as full hex")
    fun whenIFormatTheAddressAsFullHex() {
        runCatching {
            world.hexString = (world.address as AccountAddress).toString()
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I format the address as short string")
    fun whenIFormatTheAddressAsShortString() {
        runCatching {
            val addr = world.address as AccountAddress
            // Remove leading zeros for short format
            val full = addr.toString().removePrefix("0x")
            val short = full.trimStart('0').ifEmpty { "0" }
            world.hexString = "0x$short"
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I serialize the address with BCS")
    fun whenISerializeTheAddressWithBcs() {
        runCatching {
            val addr = world.address as AccountAddress
            // AccountAddress is 32 bytes - convert from hex representation
            val hex = addr.toString().removePrefix("0x").padStart(64, '0')
            world.serializedBytes = hex.hexToBytes()
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I deserialize the bytes as AccountAddress")
    fun whenIDeserializeTheBytesAsAccountAddress() {
        runCatching {
            val hex = "0x" + world.bytes!!.toHex()
            world.address = AccountAddress.fromString(hex)
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I compare the two addresses")
    fun whenICompareTheTwoAddresses() {
        // Comparison happens in the Then step
    }
    
    @When("I format it as full hex")
    fun whenIFormatItAsFullHex() {
        runCatching {
            val addr = world.address as AccountAddress
            world.hexString = addr.toString()
            // Ensure it's padded to full length with 0x prefix
            val hex = world.hexString!!.removePrefix("0x").padStart(64, '0')
            world.hexString = "0x$hex"
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I format it as short string")
    fun whenIFormatItAsShortString() {
        runCatching {
            val addr = world.address as AccountAddress
            val full = addr.toString().removePrefix("0x")
            val short = full.trimStart('0').ifEmpty { "0" }
            world.hexString = "0x$short"
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I BCS serialize the address")
    fun whenIBcsSerializeTheAddress() {
        runCatching {
            val addr = world.address as AccountAddress
            // BCS serialization of AccountAddress is just the 32 bytes
            world.serializedBytes = getAddressBytes()
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I BCS deserialize as AccountAddress")
    fun whenIBcsDeserializeAsAccountAddress() {
        runCatching {
            val hex = "0x" + world.bytes!!.toHex()
            world.address = AccountAddress.fromString(hex)
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @When("I BCS deserialize the result as AccountAddress")
    fun whenIBcsDeserializeTheResultAsAccountAddress() {
        runCatching {
            val hex = "0x" + world.serializedBytes!!.toHex()
            world.address2 = AccountAddress.fromString(hex)
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    // ============================================================
    // Then Steps
    // ============================================================
    
    // "the parsing should succeed" and "the parsing should fail" are in CommonSteps
    
    @Then("the parsing should fail with {string}")
    fun thenTheParsingShouldFailWith(errorType: String) {
        world.error shouldNotBe null
    }
    
    @Then("the parsing should fail with an invalid address error")
    fun thenTheParsingShouldFailWithInvalidAddressError() {
        world.error shouldNotBe null
    }
    
    @Then("the parsing should fail with an invalid hex error")
    fun thenTheParsingShouldFailWithInvalidHexError() {
        world.error shouldNotBe null
    }
    
    @Then("the parsing should fail with an invalid length error")
    fun thenTheParsingShouldFailWithInvalidLengthError() {
        world.error shouldNotBe null
    }
    
    @Then("the full hex should be {string}")
    fun thenTheFullHexShouldBe(expected: String) {
        val addr = world.address as AccountAddress
        val actual = addr.toString().removePrefix("0x").lowercase().padStart(64, '0')
        val exp = expected.removePrefix("0x").lowercase()
        actual shouldBe exp
    }
    
    @Then("the short string should be {string}")
    fun thenTheShortStringShouldBe(expected: String) {
        // If hexString was set by a When step, use it
        if (world.hexString != null) {
            world.hexString!!.lowercase() shouldBe expected.lowercase()
        } else {
            // Otherwise compute from address
            val addr = world.address as AccountAddress
            val full = addr.toString().removePrefix("0x")
            val short = "0x" + full.trimStart('0').ifEmpty { "0" }
            short.lowercase() shouldBe expected.lowercase()
        }
    }
    
    @Then("the bytes should be {string}")
    fun thenTheBytesShouldBe(expected: String) {
        val expectedBytes = expected.hexToBytes()
        val addrBytes = getAddressBytes()
        addrBytes shouldBe expectedBytes
    }
    
    @Then("the addresses should be equal")
    fun thenTheAddressesShouldBeEqual() {
        world.address shouldBe world.address2
    }
    
    @Then("the addresses should not be equal")
    fun thenTheAddressesShouldNotBeEqual() {
        world.address shouldNotBe world.address2
    }
    
    @Then("the address bytes should have length {int}")
    fun thenTheAddressBytesShouldHaveLength(expectedLength: Int) {
        val bytes = getAddressBytes()
        bytes.size shouldBe expectedLength
    }
    
    @Then("byte {int} should equal {int}")
    fun thenByteShouldEqual(index: Int, expectedValue: Int) {
        val bytes = getAddressBytes()
        (bytes[index].toInt() and 0xFF) shouldBe expectedValue
    }
    
    @Then("bytes {int}-{int} should all be {int}")
    fun thenBytesShouldAllBe(start: Int, end: Int, expectedValue: Int) {
        val bytes = getAddressBytes()
        for (i in start..end) {
            (bytes[i].toInt() and 0xFF) shouldBe expectedValue
        }
    }
    
    @Then("all {int} bytes should be {int}")
    fun thenAllBytesShouldBe(count: Int, expectedValue: Int) {
        val bytes = getAddressBytes()
        bytes.size shouldBe count
        bytes.all { (it.toInt() and 0xFF) == expectedValue } shouldBe true
    }
    
    // "the result should be {int} bytes" is in SerializationSteps
    
    @Then("the result should equal the original address")
    fun thenTheResultShouldEqualTheOriginalAddress() {
        val addr1 = world.address as AccountAddress
        val addr2 = world.address2 as AccountAddress
        addr1.toString().lowercase() shouldBe addr2.toString().lowercase()
    }
    
    @Then("the two addresses should be equal")
    fun thenTheTwoAddressesShouldBeEqual() {
        val addr1 = world.address as AccountAddress
        val addr2 = world.address2 as AccountAddress
        addr1.toString().lowercase() shouldBe addr2.toString().lowercase()
    }
    
    @Then("the two addresses should not be equal")
    fun thenTheTwoAddressesShouldNotBeEqual() {
        val addr1 = world.address as AccountAddress
        val addr2 = world.address2 as AccountAddress
        addr1.toString().lowercase() shouldNotBe addr2.toString().lowercase()
    }
    
    // "the serialized bytes should be {string}" is in CommonSteps
    
    // "the result should be {string}" is in CommonSteps
    
    // ============================================================
    // Special Address Steps
    // ============================================================
    
    @Given("the ZERO address constant")
    fun givenTheZeroAddressConstant() {
        runCatching {
            world.address = AccountAddress.fromString("0x0")
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @Given("the ONE address constant")
    fun givenTheOneAddressConstant() {
        runCatching {
            world.address = AccountAddress.fromString("0x1")
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @Given("the THREE address constant")
    fun givenTheThreeAddressConstant() {
        runCatching {
            world.address = AccountAddress.fromString("0x3")
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
    
    @Given("the FOUR address constant")
    fun givenTheFourAddressConstant() {
        runCatching {
            world.address = AccountAddress.fromString("0x4")
        }.onSuccess {
            world.clearError()
        }.onFailure {
            world.recordError(it)
        }
    }
}
